### ISSUE TITLE
submenu shows up far from parent on Wayland

### ISSUE THREAD

**Issue by lilydjwg** [telegramdesktop/tdesktop#24230]
### Steps to reproduce

1. maximize telegram
2. right click on the message input
3. choose to change text format

### Expected behaviour

The submenu is shown near the parent entry.

### Actual behaviour

See screenshot:

![screenshot-2022-03-22_12:42:18](https://user-images.githubusercontent.com/440661/159410627-e0fff24b-8f05-4123-970f-84e007cd2aad.png)

PS: I tried with the static binary from official site, the menu even doesn't pop up. If it were not maximized, and moved near the bottom of screen instead, this issue would also occur.

### Operating system

Arch Linux, wayfire

### Version of Telegram Desktop

3.6.1

### Installation source

Other (unofficial) source

**Comment by Aokromes** [comment]
plz try downloading official version from https://desktop.telegram.org

**Comment by lilydjwg** [comment]
Did you read my report? I tried; it was worse.

**Comment by ilya-fedin** [comment]
Would be nice to get some help with this. I don't understand how to make menus work as expected on Wayland since there's no way to get window position.

**Comment by lilydjwg** [comment]
> I don't understand how to make menus work as expected on Wayland since there's no way to get window position.

I don't know about Qt, but in Wayland protocol, you specify the position of a popup relative to its parent, and in case of reaching edges, what should be done (sliding or flipping etc), via the [xdg_positioner](https://wayland.app/protocols/xdg-shell#xdg_positioner) interface. GTK 3 has similar APIs to position a popup.

**Comment by ilya-fedin** [comment]
Qt doesn't have such APIs, it has [pre-defined xdg-positioner flags](https://github.com/qt/qtwayland/blob/dev/src/plugins/shellintegration/xdg-shell/qwaylandxdgshell.cpp#L434-L441) for all windows of type Qt::Popup (what menus are). We need [xdg_positioner::set_offset](https://wayland.app/protocols/xdg-shell#xdg_positioner:request:set_offset) I guess in order to set shadow extents, but we have no access to that API.

**Comment by lilydjwg** [comment]
Sad :cry:

**Comment by lilydjwg** [comment]
No, I'm still interested this issue to be resolved. Inactivity only means that nothing could be done currently but the problem doesn't disappear on its own.

There should be some way to keep inactive issues open if stalebot is enabled.

**Comment by ilya-fedin** [comment]
> but the problem doesn't disappear on its own.

Are you sure? My experience tells they do. Can you re-check?

**Comment by lilydjwg** [comment]
I just tried version 4.1.1 official build, it still misaligns, not as high as the above screenshot though.

**Comment by ilya-fedin** [comment]
Can you check 4.2.0? It uses newer Qt

**Comment by lilydjwg** [comment]
It got updated again? I'll update and see soon.

**Comment by lilydjwg** [comment]
Same with 4.2.0:

![图片](https://user-images.githubusercontent.com/440661/190949008-6193c2d0-fe6d-46b2-8949-8a06f6283152.png)

I have another build with Arch's Qt6 and it has exactly the same issue too.

**Comment by ilya-fedin** [comment]
This seems to be the best thing Qt can do with its design...

**Comment by lilydjwg** [comment]
I just notice that there is a warning message from Qt:

```
qt.qpa.wayland: setGrabPopup called with a parent, QtWaylandClient::QWaylandXdgSurface(0x7f70967fcf60) which does not match the current topmost grabbing popup, QtWaylandClient::QWaylandXdgSurface(0x7f704a3f23a0) According to the xdg-shell protocol, this is not allowed. The wayland QPA plugin is currently handling it by setting the parent to the topmost grabbing popup. Note, however, that this may cause positioning errors and popups closing unxpectedly because xdg-shell mandate that child popups close before parents
```

I don't  know what it means but it may help?

**Comment by ilya-fedin** [comment]
This is about some Wayland's weird thing with strict hierarchy that tdesktop can't obviously follow due to being a cross-platform application, it seems Qt does a good job here by guessing and automatically changing the parents

**Comment by Aokromes** [comment]
> No, I'm still interested this issue to be resolved. Inactivity only means that nothing could be done currently but the problem doesn't disappear on its own.
> 
> There should be some way to keep inactive issues open if stalebot is enabled.

slatebot helps us to find fixes issues witch their issues aren't linked on fix and closed on fix time.

**Comment by ilya-fedin** [comment]
This issue is now a part of #25126

**Comment by lilydjwg** [comment]
Things get worse with 4.2 with context menus **on messages** affected: they may appear high above the click position when the message is too below, but when I hover *below* the menu, items on the menu actually get highlighted, i.e. when I click empty space below the menu, trying to dismiss the menu, an item is activated instead.

The highlighted item issue has been around but this time, the menu shifts above, leaving a lot of space to accidentally trigger it.

Tested on both official build and a build with Qt6.

**Comment by ilya-fedin** [comment]
It's due to reactions I guess, the top-left point of the menu window is where the reaction bubble (way upper than the cursor), so it becomes new bottom(-right) point. I have no idea how to fix that.

**Comment by ilya-fedin** [comment]
I don't believe the problem is in the parents in tdesktop case, but you're free to prove that

**Comment by ilya-fedin** [comment]
QTBUG-87303 just got a follow-up commit that [changes constraint strategy](https://codereview.qt-project.org/c/qt/qtwayland/+/453635) on Wayland, so this can become unreproducible in Qt 6.5.0-beta4

**Comment by lilydjwg** [comment]
It's fixed in Qt 6.5, thanks!
